editor/src/components/canvas/controls/select-mode/absolute-resize-control.tsx

```diff
@@ -0,0 +1,349 @@
+import React from 'react'
+import { LayoutTargetableProp } from '../../../../core/layout/layout-helpers-new'
+import { MetadataUtils } from '../../../../core/model/element-metadata-utils'
+import { ElementInstanceMetadataMap } from '../../../../core/shared/element-template'
+import {
+  boundingRectangleArray,
+  CanvasPoint,
+  CanvasRectangle,
+  windowPoint,
+} from '../../../../core/shared/math-utils'
+import { ElementPath } from '../../../../core/shared/project-file-types'
+import { useColorTheme } from '../../../../uuiui'
+import { EditorDispatch } from '../../../editor/action-types'
+import { setResizeOptionsTargetOptions } from '../../../editor/actions/action-creators'
+import { useEditorState, useRefEditorState } from '../../../editor/store/store-hook'
+import CanvasActions from '../../canvas-actions'
+import {
+  CSSCursor,
+  DirectionAll,
+  DirectionHorizontal,
+  DirectionVertical,
+  EdgePosition,
+  resizeDragState,
+  updateResizeDragState,
+} from '../../canvas-types'
+import { getOriginalFrames } from '../../canvas-utils'
+import { windowToCanvasCoordinatesGlobal } from '../../dom-lookup'
+import { findFramesFromDOM, useMutationObserver } from '../observer-hooks'
+
+interface AbsoluteResizeControlProps {
+  selectedElements: Array<ElementPath>
+}
+
+export const AbsoluteResizeControl = React.memo<AbsoluteResizeControlProps>((props) => {
+  const controlRef = React.useRef<HTMLDivElement>(null)
+  const topLeftRef = React.useRef<HTMLDivElement>(null)
+  const topRightRef = React.useRef<HTMLDivElement>(null)
+  const bottomLeftRef = React.useRef<HTMLDivElement>(null)
+  const bottomRightRef = React.useRef<HTMLDivElement>(null)
+
+  const leftRef = React.useRef<HTMLDivElement>(null)
+  const topRef = React.useRef<HTMLDivElement>(null)
+  const rightRef = React.useRef<HTMLDivElement>(null)
+  const bottomRef = React.useRef<HTMLDivElement>(null)
+
+  const selectedElements = props.selectedElements
+  const selectedElementsRef = React.useRef(selectedElements)
+  selectedElementsRef.current = selectedElements // TODO new-canvas-controls@localSelectedViews should be atom-like so we can get a live ref to it
+
+  const allSelectedElementsAbsolute = useEditorState((store) => {
+    return (
+      selectedElementsRef.current.length > 0 &&
+      selectedElementsRef.current.every((path) => {
```

Comment:
Why is this using the ref rather than `selectedElements`?

Reply:
we should rename selectedElements to localSelectedElements and selectedElementsRef to localSelectedElementsRef.

TLDR is that the canvas controls have a local (inner) selected elements array that is updated one frame before dispatching a selection change. this was implemented so that the visual selection change is as fast as possible, and then the slow inspector rerender happens _after_ we already drew the new controls.

I added a TODO to the paper doc about this, it has to do with the `TODO new-canvas-controls@localSelectedViews` todo.